### PR TITLE
feat: add cron job

### LINKED ISSUE #157: create a cronjob to call self_destruct contract method at the end of every month
await PromptWarsMarketContract.selfDestruct(marketId!);

https://api.github.com/aufacicenta/pulsemarkets/blob/7502d455aa9430e1fa2251e28ba49a9dd20e0071/app/src/pages/api/prompt-wars/resolve.ts#L31

```typescript


    await PromptWarsMarketContract.sellResolved(marketId!);

    const marketContract = await PromptWarsMarketContract.loadFromGuestConnection(marketId!);

    const blockTimestamp = await marketContract.get_block_timestamp();
    const resolution = await marketContract.get_resolution_data();
    const ms = resolution.window - blockTimestamp;

    logger.info(`setting timeout to call the claim_fees method for market ${marketId} in ${ms} ms`);
    setTimeout(async () => {
      try {
        logger.info(`calling the claim_fees method for market ${marketId}`);
        await PromptWarsMarketContract.claimFees(marketId!);
      } catch (error) {
        logger.error(error);
      }
    }, ms);

    // @TODO create a cronjob to call self_destruct contract method at the end of every month
    // labels: 250 USDT
    // await PromptWarsMarketContract.selfDestruct(marketId!);

    response.status(200).json({ marketId });
  } catch (error) {

```

### LINKED ISSUE #202: separate these rules and error on each with a specific message
https://api.github.com/aufacicenta/pulsemarkets/blob/3d0163942f80005a1e7c70f80fff8445afa97ec5/contracts/prompt-wars/src/contract.rs#L242

```rust

            env::panic_str("ERR_SELF_DESTRUCT_WINDOW_NOT_EXPIRED");
        }

        // @TODO separate these rules and error on each with a specific message
        // labels: 100 USDT
        if !self.get_outcome_ids().is_empty()
            && !self.fees.claimed_at.is_some()
            && self.collateral_token.fee_balance > 0

```

### PR DESCRIPTION
**Pull request by Udit-takkar**
Fixes: https://github.com/aufacicenta/pulsemarkets/issues/157
Fixes: https://github.com/aufacicenta/pulsemarkets/issues/202

### CODE REVIEW

File: .github/workflows/cron-selfDestruct.yml [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+on:
+  schedule:
+    - cron: "0 0 1 * *" # This runs on the 1st day of every month at midnight
+
+jobs:
+  cron-selfDestruct:
+    env:
+      APP_URL: ${{ secrets.APP_URL }}
+      CRON_API_KEY: ${{ secrets.CRON_API_KEY }}
```
Comment (Udit-takkar) [inline .github/workflows/cron-selfDestruct.yml:11]: @netpoe you would have to add APP_URL and CRON_API_KEY in github settings.

File: app/src/pages/api/cron/selfDestruct.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
+import { NextApiRequest, NextApiResponse } from "next";
```
Comment (netpoe) [inline app/src/pages/api/cron/selfDestruct.ts:1]: Let's make the endpoint file to be: `self-destruct.ts` to keep naming conventions ok.

File: app/src/pages/api/cron/selfDestruct.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  const apiKey = _request.headers.authorization || _request.query.apiKey;
+
+  if (process.env.CRON_API_KEY !== apiKey) {
+    return response.status(401).json({ message: "Not authenticated" });
+  }
+
+  if (_request.method !== "POST") {
+    return response.status(405).json({ message: "Invalid method" });
+  }
+  try {
```
Comment (netpoe) [inline app/src/pages/api/cron/selfDestruct.ts:17]:
```suggestion
  }
  
  try {
```

File: app/src/pages/api/cron/selfDestruct.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    const marketFactory = await MarketFactoryContract.loadFromGuestConnection();
+    const marketsList = await marketFactory.get_markets_list();
+
+    await Promise.all(
+      marketsList!.map(async (marketId) => {
+        try {
+          const contract = await PromptWarsMarketContract.loadFromGuestConnection(marketId!);
+          const is_self_destruct_window_expired = await contract.is_self_destruct_window_expired();
```
Comment (netpoe) [inline app/src/pages/api/cron/selfDestruct.ts:25]:
```suggestion
          const isSelfDestructWindowExpired = await contract.is_self_destruct_window_expired();
```

**Review by netpoe — COMMENTED** [review]
@Udit-takkar this issue is a tricky one. It should work together with: https://github.com/aufacicenta/pulsemarkets/issues/202

When calling the self_destruct method, we have seen an issue in the Rust implementation. IF blocks need to be separated. Otherwise this `cron-selfDestruct` endpoint may result unreliable.

File: contracts/prompt-wars/src/contract.rs [hunk 5]
```diff
@@ -239,13 +239,18 @@ impl Market {
...
+        }
+
+        if !self.fees.claimed_at.is_some() {
             env::panic_str("ERR_SELF_DESTRUCT_FEES_UNCLAIMED");
         }
 
+        if self.collateral_token.fee_balance > 0 {
+            env::panic_str("ERR_COLLATERAL_TOKEN_FEE_BALANCE_GREATER_THAN_ZERO");
+        }
+
```
Comment (Udit-takkar) [inline contracts/prompt-wars/src/contract.rs:253]: @netpoe done. is this fine?

Comment (netpoe) [reply contracts/prompt-wars/src/contract.rs:253]: Let me try it out. I'll let you know. Thx.

**Comment by Udit-takkar** [comment]
@netpoe any update or feedback?

**Comment by netpoe** [comment]
> @netpoe any update or feedback?

@Udit-takkar I merged your latest changes to this one and 3 tests fail: 

```
failures:

---- tests::tests::err_self_destruct_unresolved stdout ----
note: panic did not contain expected string
      panic message: `"called `Result::unwrap()` on an `Err` value: HostError(GuestPanic { panic_msg: \"ERR_NO_OUTCOME_IDS\" })"`,
 expected substring: `"ERR_SELF_DESTRUCT_FEES_UNCLAIMED"`

failures:
    tests::tests::err_self_destruct_unresolved
    tests::tests::self_destruct_resolved_no_players
    tests::tests::sell_resolved

test result: FAILED. 10 passed; 3 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.01s

error: test failed, to rerun pass `--lib`
```

Would you like to attempt and fix them? I can add another 100 USDT to the bounty for extending `tests.rs` to account for these additional errors introduced in `self_destruct`.

**Comment by Udit-takkar** [comment]
@netpoe okay fixed all the tests

Screenshot 2023-10-27 at 12 06 11 AM

**Comment by netpoe** [comment]
@Udit-takkar I cannot push to your branch so please merge this branch: `aufacicenta:udit-takkar` into yours: `Udit-takkar:feat/cron-jobs`. I made some fixes and refactors to account for this cronjob feature.

**Comment by Udit-takkar** [comment]
@netpoe done.

**Comment by netpoe** [comment]
> @netpoe done.

Tests pass.
Let me update the Factory contracts, deploy and test the cronjob. Give me a few days. Thx for your patience.

**Comment by netpoe** [comment]
@Udit-takkar https://explorer.testnet.near.org/transactions/41DCiMsUxHG4Z8De9HEXo3Kdxui5zCGdxtDTWF7PGgQT 
I've deployed the market-factory with the self_destruct updates. We need to wait for 7 days to test the cronjob on the most recent games.

I'll go ahead and merge this to avoid conflicts on potential changes. 

I have your payment in mind. Don't worry.

**Comment by Udit-takkar** [comment]
@netpoe ok.

 just confirming did you add CRON_API_KEY in secrets and .env file and APP_URL in secrets ?

**Comment by netpoe** [comment]
> @netpoe ok.
> 
> just confirming did you add CRON_API_KEY in secrets and .env file and APP_URL in secrets ?

Not yet. But I'll keep an eye on the GH Actions logs.

**Comment by Udit-takkar** [comment]
> @Udit-takkar https://explorer.testnet.near.org/transactions/41DCiMsUxHG4Z8De9HEXo3Kdxui5zCGdxtDTWF7PGgQT I've deployed the market-factory with the self_destruct updates. We need to wait for 7 days to test the cronjob on the most recent games.
> 
> I'll go ahead and merge this to avoid conflicts on potential changes.
> 
> I have your payment in mind. Don't worry.

@netpoe any update on this?

**Comment by netpoe** [comment]
> > @Udit-takkar https://explorer.testnet.near.org/transactions/41DCiMsUxHG4Z8De9HEXo3Kdxui5zCGdxtDTWF7PGgQT I've deployed the market-factory with the self_destruct updates. We need to wait for 7 days to test the cronjob on the most recent games.
> > I'll go ahead and merge this to avoid conflicts on potential changes.
> > I have your payment in mind. Don't worry.
> 
> @netpoe any update on this?

I tested by calling the ´self-destruct´ endpoint manually, before activating the cron job. I know the cron-job will work, but I need to make sure that the contracts are actually self-destroyed, which in this case, they weren't. Even with the latest updates we did to the Rust logic.

**Comment by Udit-takkar** [comment]
> I tested by calling the ´self-destruct´ endpoint manually, before activating the cron job. I know the cron-job will work, but I need to make sure that the contracts are actually self-destroyed, which in this case, they weren't. Even with the latest updates we did to the Rust logic.

 Okay do you see anything wrong in app/src/pages/api/cron/self-destruct.ts ? or any other error that were logged while testing

**Comment by netpoe** [comment]
> > I tested by calling the ´self-destruct´ endpoint manually, before activating the cron job. I know the cron-job will work, but I need to make sure that the contracts are actually self-destroyed, which in this case, they weren't. Even with the latest updates we did to the Rust logic.
> 
> Okay do you see anything wrong in app/src/pages/api/cron/self-destruct.ts ? or any other error that were logged while testing

Yes. While testing, I get errors of contracts with the old ´self_destruct´ logic. Since we are moving away from NEAR, I prefer not to invest much time into these Rust issues. Otherwise, I'd need to deploy a new NEAR market factory, etc.

**Pull request merged**